dht/range_streamer.cc

```diff
@@ -288,22 +286,24 @@ future<> range_streamer::stream_async() {
                             _nr_total_ranges - remaining, _nr_total_ranges, _reason, percentage);
                 };
                 try {
-                    for (auto it = range_vec.begin(); it < range_vec.end();) {
-                        ranges_to_stream.push_back(*it);
-                        it = range_vec.erase(it);
+                  if (range_vec.size() <= nr_ranges_per_stream_plan) {
+                    do_streaming(std::move(range_vec));
+                  } else {
+                    dht::token_range_vector ranges_to_stream;
+                    for (const auto& r : range_vec) {
+                        ranges_to_stream.push_back(r);
                         if (ranges_to_stream.size() < nr_ranges_per_stream_plan) {
                             continue;
                         } else {
-                            do_streaming();
+                            do_streaming(std::move(ranges_to_stream));
```

Comment:
The ranges_to_stream is moved here, but we are accessing it again at line 294 in the loop.

Reply:
Right. The content of the vector is moved to `do_streaming` and it becomes empty.
The access at line 294 when we loop checks the vector after inserting more ranges to it at line 292 so I see no problem with that.

Reply:
IIRC, a moved from vector is valid but unspecified. It is not guaranteed to be empty. 

E.g., see https://stackoverflow.com/questions/17730689/is-a-moved-from-vector-always-empty

Reply:
@avikivity How do you think? 

How about this?
```
do_streaming(std::move(ranges_to_stream));
ranges_to_stream = {};
```

Reply:
you're right.
https://en.cppreference.com/w/cpp/language/move_constructor even explicitly mentions thr std::vector case.

Reply:
I think that `.clear()` would be clearer than `= {}` :)

Reply:
fixed in v3

Reply:
I am not sure why .clear is better.  Calling .clear after the vector is moved is still accessing  a removed from object (which we should avoid in general). The removed from vector is still in a unspecified state (Yes,  clear makes .size() return zero and invalidate references and so do). 

In stead

ranges_to_stream = {};

makes it clear, we assign a new object to ranges_to_stream, we do not care about the old one anymore (which is removed from vector).

@avikivity ?

Reply:
@asias there should be no provlem accessing the moved-from vector. The standard says it ends up in an unspecified, but valid, state.  So it's not guranteed to be empty, but it may be accessed for e.g. clear().  If it would leave the capacity of the moved-frim vector by moving all the objects from it, then clear will save an allocation. Assiging an empty vector will guarantee allocation (though in practice the moved vector does lose its capacity and allocation will happen if we insert more tokens in the loop)

Reply:
IMHO, it is better to follow the simple rule: never touch a moved from object than playing games to save an allocation.
The code is on the control plane anyway.

Reply:
This rule is an overkill.  Move semantics require that rhe moved-from object will remain valid, otherwise destroying it will easily end up with undefined behavior since the destructor is always called on it before it's destroyed.
I agree that if its state is unspecified the code shouldn't make any assumptions about its state post-move, but I see no provlem in getting it back into a specified state.

Consider a moved-from optional, which is used quite extensively in our code. You know it still evaluates to true as it is still engaged, holding a moved-from object, any in many cases, we reset this moved-from optional to get it into the disengaged state.

Reply:
Also, @asias, assigning the moved-from object with a default-constructed object is touching it, as it executes the move-assignment operator which typically destroys the assigned object before move-constructing it from the rvalue object.

Reply:
Benny, it's a good point that  default constructor is touching it too. It is fair enough. It's up to the maintainers to decide.

Reply:
My preference is to use std::exchange() when you want to ensure the object is in a known state after moving it.

Reply:
I don't think that works for exception safety as std::exchange takes effect before the rvalue is passed to the called function, and if throws, it's lost.

I'm going to send another version though with improved exception safety that should be less contentious around clearing the ranges vector past move